sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java

```diff
@@ -136,77 +195,95 @@ public Write withDatabaseId(String databaseId) {
       return toBuilder().setDatabaseId(databaseId).build();
     }
 
+    @VisibleForTesting
+    Write withServiceFactory(ServiceFactory<Spanner, SpannerOptions> serviceFactory) {
+      return toBuilder().setServiceFactory(serviceFactory).build();
+    }
+
     @Override
-    public PDone expand(PCollection<Mutation> input) {
-      input.apply("Write mutations to Spanner",
-          ParDo.of(new SpannerWriterFn(
-              getInstanceId(), getDatabaseId(), SPANNER_MUTATIONS_PER_COMMIT_LIMIT)));
+    public void validate(PipelineOptions options) {
+      checkNotNull(
+          getInstanceId(),
+          "SpannerIO.write() requires instance id to be set with withInstanceId method");
+      checkNotNull(
+          getDatabaseId(),
+          "SpannerIO.write() requires database id to be set with withDatabaseId method");
+    }
 
+    @Override
+    public PDone expand(PCollection<Mutation> input) {
+      input.apply("Write mutations to Cloud Spanner", ParDo.of(new SpannerWriteFn(this)));
       return PDone.in(input.getPipeline());
     }
 
     @Override
     public void populateDisplayData(DisplayData.Builder builder) {
       super.populateDisplayData(builder);
       builder
-          .addIfNotNull(DisplayData.item("instanceId", getInstanceId())
-              .withLabel("Output Instance"))
-          .addIfNotNull(DisplayData.item("databaseId", getDatabaseId())
-              .withLabel("Output Database"));
+          .addIfNotNull(DisplayData.item("projectId", getProjectId()).withLabel("Output Project"))
+          .addIfNotNull(
+              DisplayData.item("instanceId", getInstanceId()).withLabel("Output Instance"))
+          .addIfNotNull(
+              DisplayData.item("databaseId", getDatabaseId()).withLabel("Output Database"))
+          .addIfNotNull(
```

Comment:
1. this can't be null, it's a `long`.
2. Use `addIfNotDefault` so it doesn't show up when left to default settings. https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/display/DisplayData.java#L212